src/include/rho/AllocationTable.hpp

```diff
@@ -0,0 +1,326 @@
+       */
+      bool isSuperblock() {
+        return (m_data & s_superblock_flag) == s_superblock_flag;
+      }
+
+      /**
+       * Returns the data pointer as a void*.
+       */
+      void* asPointer() {
+        return reinterpret_cast<void*>(dataPointer());
+      }
+
+      /**
+       * Returns the data pointer as a superblock pointer.
+       */
+      rho::AllocatorSuperblock* asSuperblock() {
+        return reinterpret_cast<rho::AllocatorSuperblock*>(dataPointer());
+      }
+
+    private:
+      friend AllocationTable;
+
+      /** Constant key for an empty hash bucket. */
+      static constexpr uintptr_t s_empty_key = 0;
+
+      /** Constant key for a deleted hash bucket. */
+      static constexpr uintptr_t s_deleted_key = UINTPTR_MAX;
+
+      /** The flag bit indicating the first bucket for an allocation. */
+      static constexpr int s_first_flag = 1;
+
+      /** The flag bit indicating that the allocation is a superblock. */
+      static constexpr int s_superblock_flag = 2;
+
+      uintptr_t m_data;
+
+      /** The size in bytes for small objects, or the 2-log for large objects. */
+      unsigned m_size_log2;
+
+      /** Masks away the pointer bits from the data field. */
+      uintptr_t dataPointer() {
+        return m_data & ~uintptr_t{3};
+      }
+
+      /** Returns true if this hash bucket represents a deleted allocation. */
+      bool isDeleted() {
+        return m_data == s_deleted_key;
+      }
+
+      /** Returns true if this hash bucket is unused. */
+      bool isEmpty() {
+        return m_data == s_empty_key;
+      }
+
+      /**
+       * Returns true if this allocation bucket is the first entry for the
+       * corresponding allocation in the allocation table.
+       */
+      bool isFirst() {
+        return (m_data & s_first_flag) == s_first_flag;
+      }
+
+      /**
+       * Returns true if the argument pointer points inside this allocation.
+       */
+      bool isInternalPointer(uintptr_t pointer) {
+        uintptr_t data = dataPointer();
+        return data <= pointer && (data + (1L << m_size_log2)) > pointer;
+      }
+
+    };
+
+    /** The low pointer bits are shifted out in the hash function. */
+    static constexpr unsigned LOW_BITS = 19;
+
+    /** Additional entries that can be inserted before resize. */
+    unsigned m_capacity;
+
+    unsigned m_num_bits;
+    unsigned m_num_buckets;
+
+    /** Mask deciding which bits of a hash key are used to index the hash table. */
+    unsigned m_hash_mask;
+
+    Allocation* m_buckets = nullptr;
+
+    // Counters logging collision statistics:
+    unsigned m_num_insert_collisions = 0;
+    unsigned m_num_erase_collisions = 0;
+    unsigned m_num_search_collisions = 0;
+
+    /**
+     * Unlike most other member functions this one takes a table pointer,
+     * because it is used to tentatively insert new nodes into a temporary
+     * resized table.
+     *
+     * @return true if the insert succeeded.
```

Comment:
With the probe sequence that you're using, insertion should only fail if the table is completely full. Make it a precondition that the table has enough space, and you can remove the failure case, which will make things simpler.